mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -2176,22 +2197,25 @@ struct GridwiseAttentionAccelRewritePattern
               toLDSRegBufferQ, ldsByteBufferQ, "n", gemm0kpack,
               gemm0KpacksPerBlock, gemm0NPerBlock, blockSize, gridSize,
               bidGridOrder, gemm0BidGridLengths, forceUnroll, rewriter,
-              *accelEmitterPtrGemm0, ldsLayoutCfgNG0);
+              *accelEmitterPtrGemm0, ldsLayoutCfgNG0, addBarrierFirstGemm);
           if (failed(statusLoadQ)) {
             return failure();
           }
           ldsTileBufferQ =
               viewBufferAs(rewriter, ldsByteBufferQ,
                            vectorTypeOrSelf(elemTypeQ, gemm0kpack));
         }
+        // if we added a barrier in the previous block (load Q), there's no need
+        // to add it again here.
+        bool fallbackBarrier = addBarrierFirstGemm && gemm0K == gemm0KPerBlock;
```

Comment:
is this working as expected ?

I expect it to be more like `fallbackbarrier = !addBarrierFirstGemm`